src/iop/demosaic.c

```diff
@@ -5023,6 +5024,12 @@ void reload_defaults(dt_iop_module_t *module)
   {
     module->default_enabled = 0;
   }
+
+  dt_iop_demosaic_gui_data_t *g = (dt_iop_demosaic_gui_data_t *)module->gui_data;
```

Comment:
gui stuff should not be part of reload_defaults(), it should be in gui_update()

Reply:
This is not completely correct. reload_defaults gets called both in headless and gui mode and if any settings (for example defaults) should be changed on a per-image basis, this should happen here.

The problem is that upon switching to darkroom, reload_defaults is called with a newly created gui widget that has not yet been added to the right panel. This means it behaves slightly differently from when it later gets called if the image is changed from the filmstrip. Maybe this is because it has not been realized and mapped yet, or maybe a `show` is called when it is added, so that any `hide`s are undone (which is a common per-image action if some widget is not applicable).
Because of this, some of the once-per-image actions that should be in reload_defaults are put in update_gui, so that they actually work.

I'm working on a follow up to #6240 that hopefully succeeds in removing this last issue in init/reload/update separation. But since it will mess with initialisation order, I'm worried about corner cases and will need a lot of testing to cover all of them.

Reply:
@dterrahe : ok I though that we already did a gui_upadet() after reload_defaults() if a GUI was set. So yes I'm looking forward to your next clean-up, this will certainly clarify the responsibility of each hook.

Reply:
We _do_, so that's why putting everything in gui_update works. But sometimes the per-image changes are quite substantive (completely rebuilding comboboxes based on long lists of presets and maybe having to query databases etc) so doing them for every single gui_update would just not be efficient and it shouldn't be necessary. So the intuition would be to put that in reload_defaults (in my interpretation this just doesn't have gui_ in the name because it also gets called for non-gui). But if you do that, at the moment some things subtly break (but only on first switching to darkroom). While others work. So following intuition can be frustrating/time-wasting/bug-causing.

I'm hoping I can make the intuition just work. The fact that then some _current_ implementations can maybe also be optimised a bit is very secondary to the goal of making people waste less time in future on trying to understand all this. And fixing bugs.